### PR TITLE
fix(forksafe): better after-in-child fork support

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

On older versions of Python that do not implement `os.register_at_fork` we wrap around `os.fork` instead of using `threading._after_fork` to run our after-in-child hook as soon as possible to ensure a proper state clean-up and re-generation.

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

**Review by brettlangdon — DISMISSED** [review]

**Review by Kyle-Verhoog — COMMENTED** [review]
Can we elaborate further in the PR why using `_after_fork` was broken or not working?

**Comment by majorgreys** [comment]
A test for where `_after_fork` isn't enough would be good but would also be fine with detailed documentation for this patching.

**Comment by P403n1x87** [comment]
I am thinking of supporting both the current and the proposed way of executing code after fork. It seems that `threading._after_fork` is just fine in most cases, except when starting a new thread within the registered hooks. In those cases we could just execute the hooks by patching the `os.fork` method, which seems to work in this case. Ideally we'd want a way to detect, e.g., `celery` or other integrations that might require the alternative approach and set a flag that can be used to determine when to execute the hooks.

**Review by Kyle-Verhoog — COMMENTED** [review]
would still like to see more context about what the issues were and why/how this fixes them

**Comment by brettlangdon** [comment]
> would still like to see more context about what the issues were and why/how this fixes them


> It seems that threading._after_fork is just fine in most cases, except when starting a new thread within the registered hooks

**Comment by Kyle-Verhoog** [comment]
@brettlangdon yeah I think I'm more curious about the specifics of why it doesn't work

**Comment by P403n1x87** [comment]
> @brettlangdon yeah I think I'm more curious about the specifics of why it doesn't work

The fact that sometimes this causes a SIGSEGV and other times a Python exception on an unlocked lock makes me think that this is just a race condition within the interpreter, which doesn't like when threads are started while it is still resetting itself right soon after fork. That's the best guess that I can offer at the moment 🙁

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 1]
```diff
@@ -81,7 +80,7 @@ class RuntimeWorker(periodic.PeriodicService):
 
     enabled = False
     _instance = None  # type: ClassVar[Optional[RuntimeWorker]]
-    _lock = RLock()
+    _lock = forksafe.Lock()
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/internal/runtime/runtime_metrics.py:83]: Yes, there shouldn't be the need of an `RLock` here.

File: tests/contrib/celery/test_integration.py [hunk 2]
```diff
@@ -766,3 +770,25 @@ def fn_task():
             run_span = traces[0][0]
 
         assert run_span.trace_id == 12345
+
+    def test_thread_in_fork(self):
```
Comment (brettlangdon) [inline tests/contrib/celery/test_integration.py:774]: docstring here would be nice

File: ddtrace/contrib/celery/patch.py [hunk 3]
```diff
@@ -9,6 +10,9 @@
 from .constants import WORKER_SERVICE
 
 
+forksafe._soft = True
```
Comment (brettlangdon) [inline ddtrace/contrib/celery/patch.py:13]: comment here would be nice

**Review by brettlangdon — COMMENTED** [review]
these changes seem fine, I think my current concern is when celery is available but a worker is not running, what is the side effect on other integrations (or internal code) that would then have their after fork callbacks moved?

I wonder if we need to be more selective on when we enable `forksafe._soft`, or else should that just be the default and we stop using `threading._after_fork` ?

**Review by brettlangdon — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]
:+1: cautiously approving. If this issue comes up again we should rethink the approach

**Pull request merged**